src/System.Linq.Expressions/tests/StackSpillerTests.cs

```diff
@@ -376,78 +377,797 @@ public static void Spill_LabelGoto()
+
+                  // return v[0]
+                  IL_0016: ldloca.s   V_2
+                  IL_0018: ldc.i4.0   
+                  IL_0019: call       instance int32 valuetype [System.Linq.Expressions.Tests]System.Linq.Expressions.Tests.StackSpillerTests+ValueVector::get_Item(int32)
+                  IL_001e: ret        
+                }");
+
+            var f = e.Compile();
+
+            Assert.Equal(42, f());
         }
 
         [Fact]
-        public static void Spill_NotRefInstance_Index()
+        public static void Spill_RefInstance_Index()
         {
-            // See https://github.com/dotnet/corefx/issues/11740 for documented limitation
+            var v = Expression.Parameter(typeof(ValueBar));
+            var item = typeof(ValueBar).GetProperty("Item");
+            var x = Expression.Constant(42);
 
-            var v = Expression.Constant(new ValueVector());
-            var item = typeof(ValueVector).GetProperty("Item");
-            var i = Expression.Constant(0);
+            var e =
+                Expression.Lambda<Func<int>>(
+                    Expression.Block(
+                        new[] { v },
+                        Expression.Property(v, item, Spill(x))
+                    )
+                );
+
+            e.VerifyIL(@"
+               .method int32 ::lambda_method(class [System.Linq.Expressions]System.Runtime.CompilerServices.Closure)
+                {
+                  .maxstack 3
+                  .locals init (
+                    [0] valuetype [System.Linq.Expressions.Tests]System.Linq.Expressions.Tests.StackSpillerTests+ValueBar&,
+                    [1] int32,
+                    [2] valuetype [System.Linq.Expressions.Tests]System.Linq.Expressions.Tests.StackSpillerTests+ValueBar,
+                    [3] int32
+                  )
+
+                  // V0 = ref v                  // spill reference to v [V2] into V0
+                  IL_0000: ldloca.s   V_2
+                  IL_0002: stloc.0    
+
+                  // try { 42 } finally { }
+                  .try
+                  {
+                    IL_0003: ldc.i4.s   42
+                    IL_0005: stloc.3    
+                    IL_0006: leave      IL_000c
+                  }
+                  finally
+                  {
+                    IL_000b: endfinally 
+                  }
+                  IL_000c: ldloc.3    
+                  IL_000d: stloc.1    
+
+                  // return v[try { 42 }]        // using spilled reference in V0
+                  IL_000e: ldloc.0    
+                  IL_000f: ldloc.1    
+                  IL_0010: call       instance int32 valuetype [System.Linq.Expressions.Tests]System.Linq.Expressions.Tests.StackSpillerTests+ValueBar::get_Item(int32)
+                  IL_0015: ret        
+                }");
+
+            var f = e.Compile();
```

Comment:
Also curious - can this work with interpreter or we will have more restricted functionality there?